### PR TITLE
Implement feature to search jobs with a sp- and doc-integrated filter.

### PR DESCRIPTION
**Pull request by csadorf**
## Description

This patch enables the search for jobs with a statepoint- and document-
combined filter. To specify whether a key is a statepoint- or a
document-key use prefixes 'sp.' and 'doc.' respectively. No prefix is
equivalent to the 'sp.' prefix.

This patch is backwards compatible with the exception that the index
scheme was slightly modified.

Related to issue glotzerlab/signac-dashboard#25 .

## Minimal example

This patch enables queries such as this one:

```python
>>> project.find_jobs({'foo': 0, 'doc.bar': True})
signac.contrib.project.JobsCursor({'project': 'project', 'filter': '{'sp.foo': 0, 'doc.bar': True}'})
>>> # equivalent to:
>>> project.find_jobs({'foo': 0}, {'bar': True})
signac.contrib.project.JobsCursor({'project': 'project', 'filter': '{'sp.foo': 0, 'doc.bar': True}'})
```
Equivalently on the command line:
```bash
$ signac find foo 0 doc.bar true
```
All keys without a prefix are automatically interpreted as having the `sp.` prefix.

## Motivation and Context

It is difficult to impossible to search jobs by filter and document-filter encoded by a simple string. That makes it impossible for example to search properly using signac-dashboard. This patch enables searching with one combined filter, while searching with two filters is still possible

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

## Tasks

- [ ] Add warning about the use of `sp` and `doc` as root-level keys in documents.

### CODE REVIEW

**Comment by vyasr** [comment]
I updated your note to point to what I think is the relevant signac-dashboard issue, but please correct it if that's not the one you wanted.

**Comment by csadorf** [comment]
> I updated your note to point to what I think is the relevant signac-dashboard issue, but please correct it if that's not the one you wanted.

Thx, I've edited it slightly to say "related to" instead of "resolves", because dashboard might still need a minor update to work with this properly (at least we would need to test that).

File: signac/contrib/filterparse.py [hunk 1]
```diff
@@ -91,14 +92,65 @@ def parse_filter_arg(args, file=sys.stderr):
...
-                value = args[i+1]
-            else:
-                value = None
-            q.update(_parse_simple(key, value))
+        q = dict(parse_simple(args))
+
         return _with_message(q, file)
+
+
+def parse_simple(tokens):
```
Comment (vyasr) [inline signac/contrib/filterparse.py]: I would prefer to keep this "private" by renaming to `_parse_simple` and renaming the current `_parse_simple` to `_parse_simple_single` or so.

File: signac/contrib/filterparse.py [hunk 2]
```diff
@@ -91,14 +92,65 @@ def parse_filter_arg(args, file=sys.stderr):
...
+        if key in ('$and', '$or'):
+            if isinstance(value, list) or isinstance(value, tuple):
+                yield key, [dict(_add_prefix(item.items(), prefix)) for item in value]
+            else:
+                raise ValueError(
+                    "The argument to a logical operator must be a sequence (e.g. a list)!")
+        elif '.' in key and key.split('.', 1)[0] in ('sp', 'doc'):
+            yield key, value
+        elif key in ('sp', 'doc'):
+            yield key, value
```
Comment (vyasr) [inline signac/contrib/filterparse.py]: What is the purpose of this branch?

On a related note, should we explicitly disallow `sp` and `doc` as statepoint/document keys in addition to dots in keys?

File: signac/contrib/project.py [hunk 3]
```diff
@@ -1306,7 +1295,7 @@ def _read_cache(self):
             return cache
 
     def index(self, formats=None, depth=0,
-              skip_errors=False, include_job_document=True):
+              skip_errors=False, include_job_document=True, **kwargs):
```
Comment (vyasr) [inline signac/contrib/project.py]: Do we actually need to support arbitrary kwargs here?

**Review by vyasr — CHANGES_REQUESTED** [review]

File: signac/contrib/filterparse.py (hunk 1 above)
Comment (csadorf) [reply signac/contrib/filterparse.py]: I'm ok with the name change, however, I was actually hoping to expose this function publicly, to make it easier to implement custom user scripts like this:
```python
import click

@click.command()
@click.argument('filter', nargs=-1)
def cli(filter):
    jobs = project.find_jobs(parse_simple(filter))

cli()
```

File: signac/contrib/filterparse.py (hunk 2 above)
Comment (csadorf) [reply signac/contrib/filterparse.py]: For these cases, we need to ensure to __not__ add 'sp' or 'doc' as prefixes, for instance with a filter that is expressed like this: `{"sp": {"foo": 0}}`.

Concerning your second point, `sp` and `doc` are actually valid keys, you would need to "escape" them like this: ``{"sp.sp.foo": 0}``. However, I agree that we should probably discourage that, possibly even with warnings, and make sure to explicitly test the correctness of the behavior in that case.

File: signac/contrib/project.py (hunk 3 above)
Comment (csadorf) [reply signac/contrib/project.py]: No, that is an error. I was trying to introduce backwards compatibility here, and I think we should probably still do that in the long run, but as of right now, the `kwargs` are unnecessary.

**Review by csadorf — COMMENTED** [review]
Thx for your review @vyasr!

File: signac/contrib/filterparse.py (hunk 1 above)
Comment (bdice) [reply signac/contrib/filterparse.py]: I support making this a public method / official API - if it's not public then `signac-dashboard` has to rely on a private method.

Comment (vyasr) [reply signac/contrib/filterparse.py]: Got it, in that case I support making it part of the API as well. I still like renaming the `_parse_simple` method to indicate that it operates on one filter component at a time, but we don't have to change the `parse_simple` method.

File: signac/contrib/filterparse.py (hunk 2 above)
Comment (vyasr) [reply signac/contrib/filterparse.py]: I didn't consider that using `sp` and `doc` as root-level keys in a standard query format was an option, but in hindsight that makes sense. I support discouraging the usage of those as keys to future-proof ourselves against future changes where that might cause problems, though.

**Comment by csadorf** [comment]
I'm going to close this until we work on 1.3.

**Comment by vyasr** [comment]
@vishav1771 please comment on here and let me know if you're OK with switching to work on this for now (regarding my comments from #309).

**Comment by vishav1771** [comment]
@vyasr That's okay. 👍

**Comment by vishav1771** [comment]
This is moved to #332 .

**Pull request closed without merging**